# Анализ книжной базы данных

## Исследование данных о книгах, авторах, издательствах, оценках и обзорах

**Студент:** Оксана Петрова  
**Проект:** SQL-анализ данных книжного сервиса

---

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import text, create_engine

# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{user}:{pwd}@{host}:{port}/{db}'.format(**db_config)

# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})



In [2]:
# чтобы выполнить SQL-запрос, используем Pandas
books = '''SELECT * FROM books LIMIT 5'''
pd.io.sql.read_sql(books, con = engine)

,book_id,author_id,title,num_pages,publication_date,publisher_id
0,1,546,'Salem's Lot,594,2005-11-01,93
1,2,465,1 000 Places to See Before You Die,992,2003-05-22,336
2,3,407,13 Little Blue Envelopes (Little Blue Envelope...,322,2010-12-21,135
3,4,82,1491: New Revelations of the Americas Before C...,541,2006-10-10,309
4,5,125,1776,386,2006-07-04,268


In [3]:
# чтобы выполнить SQL-запрос, используем Pandas
authors = '''SELECT * FROM authors LIMIT 5'''
pd.io.sql.read_sql(authors, con = engine)

,author_id,author
0,1,A.S. Byatt
1,2,Aesop/Laura Harris/Laura Gibbs
2,3,Agatha Christie
3,4,Alan Brennert
4,5,Alan Moore/David Lloyd


In [4]:
# чтобы выполнить SQL-запрос, используем Pandas
publishers = '''SELECT * FROM publishers LIMIT 5'''
pd.io.sql.read_sql(publishers, con = engine)

,publisher_id,publisher
0,1,Ace
1,2,Ace Book
2,3,Ace Books
3,4,Ace Hardcover
4,5,Addison Wesley Publishing Company


In [5]:
# чтобы выполнить SQL-запрос, используем Pandas
ratings = '''SELECT * FROM ratings LIMIT 5'''
pd.io.sql.read_sql(ratings, con = engine)

,rating_id,book_id,username,rating
0,1,1,ryanfranco,4
1,2,1,grantpatricia,2
2,3,1,brandtandrea,5
3,4,2,lorichen,3
4,5,2,mariokeller,2


In [6]:
# чтобы выполнить SQL-запрос, используем Pandas
reviews = '''SELECT * FROM reviews LIMIT 5'''
pd.io.sql.read_sql(reviews, con = engine)

,review_id,book_id,username,text
0,1,1,brandtandrea,Mention society tell send professor analysis. ...
1,2,1,ryanfranco,Foot glass pretty audience hit themselves. Amo...
2,3,2,lorichen,Listen treat keep worry. Miss husband tax but ...
3,4,3,johnsonamanda,Finally month interesting blue could nature cu...
4,5,3,scotttamara,Nation purpose heavy give wait song will. List...


### Вывод: 5 таблиц в них есть связи по разным id, название столбцов нормального формата, дата в формате дата.

## - Посчитайте, сколько книг вышло после 1 января 2000 года;

In [7]:
# вычислим ко-во кгин
books_are_out = '''
SELECT COUNT(publication_date)
FROM books
WHERE publication_date >= '2000-01-01'
'''
books_are_out = pd.io.sql.read_sql(books_are_out, con = engine)
books_are_out

,count
0,821


### Вывод:
- книг вышло после 1 января 2000 года: 821

##  - Для каждой книги посчитайте количество обзоров и среднюю оценку;

In [8]:
# вычислим ко-во обзоров и среднюю оценку, отсортируем сначало по кол-ву обзоров
number_of_reviews = '''
SELECT books.title AS book_title,
    AVG (ratings.rating) AS avg_rating,
    COUNT (DISTINCT reviews.review_id) AS cnt_review
FROM books 
    INNER JOIN ratings ON ratings.book_id = books.book_id
    LEFT JOIN reviews ON reviews.book_id = books.book_id
GROUP BY books.book_id
ORDER BY cnt_review DESC, avg_rating DESC
'''

number_of_reviews = pd.io.sql.read_sql(number_of_reviews, con = engine)
number_of_reviews

,book_title,avg_rating,cnt_review
0,Twilight (Twilight #1),3.662500,7
1,Harry Potter and the Prisoner of Azkaban (Harr...,4.414634,6
2,Harry Potter and the Chamber of Secrets (Harry...,4.287500,6
3,The Book Thief,4.264151,6
4,The Glass Castle,4.206897,6
...,...,...,...
995,Disney's Beauty and the Beast (A Little Golden...,4.000000,0
996,Leonardo's Notebooks,4.000000,0
997,Essential Tales and Poems,4.000000,0
998,Anne Rice's The Vampire Lestat: A Graphic Novel,3.666667,0


In [9]:
# вычислим ко-во обзоров и среднюю оценку, отсортируем сначало по средней оценке
number_of_reviews2 = '''
SELECT books.title AS book_title,
    AVG (ratings.rating) AS avg_rating,
    COUNT (DISTINCT reviews.review_id) AS cnt_review
FROM books 
    INNER JOIN ratings ON ratings.book_id = books.book_id
    LEFT JOIN reviews ON reviews.book_id = books.book_id
GROUP BY books.book_id
ORDER BY avg_rating DESC, cnt_review DESC
'''

number_of_reviews2 = pd.io.sql.read_sql(number_of_reviews2, con = engine)
number_of_reviews2

,book_title,avg_rating,cnt_review
0,A Dirty Job (Grim Reaper #1),5.00,4
1,School's Out—Forever (Maximum Ride #2),5.00,3
2,Moneyball: The Art of Winning an Unfair Game,5.00,3
3,Arrows of the Queen (Heralds of Valdemar #1),5.00,2
4,Wherever You Go There You Are: Mindfulness Me...,5.00,2
...,...,...,...
995,The World Is Flat: A Brief History of the Twen...,2.25,3
996,Drowning Ruth,2.00,3
997,His Excellency: George Washington,2.00,2
998,Junky,2.00,2


### Вывод:
- если смотреть по кол-ву обзоров на первом месте по обзорам книга Twilight (Twilight #1) - Сумерки
- если смотреть по среднему рейтингу то на  первом месте книга A Dirty Job (Grim Reaper #1) - Грязная работенка (Мрачный жнец #1)

##   - Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;

In [10]:
# вычислим издательство 
books_thicker_50 = '''
SELECT publisher AS publisher_name,
    COUNT (books.book_id) AS cnt_books
FROM books 
    INNER JOIN publishers ON books.publisher_id = publishers.publisher_id
WHERE books.num_pages > 50
GROUP  BY publisher
ORDER  BY cnt_books DESC
LIMIT 10
'''

books_thicker_50 = pd.io.sql.read_sql(books_thicker_50, con = engine)
books_thicker_50

,publisher_name,cnt_books
0,Penguin Books,42
1,Vintage,31
2,Grand Central Publishing,25
3,Penguin Classics,24
4,Ballantine Books,19
5,Bantam,19
6,Berkley,17
7,St. Martin's Press,14
8,Berkley Books,14
9,William Morrow Paperbacks,13


### Вывод:
- издательство, которое выпустило наибольшее число книг толще 50 страниц - Penguin Books

##    - Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;

In [11]:
# вычислим автора с самой высокой средней оценкой
author_highest = '''
SELECT authors.author AS author_name,
AVG(r.average_rating) as avg_rating
FROM books 
INNER JOIN (SELECT book_id, COUNT(rating) AS cnt,
             AVG(rating) AS average_rating
      FROM ratings
      GROUP BY book_id) AS r
ON r.book_id = books.book_id
INNER JOIN authors ON authors.author_id = books.author_id
WHERE r.cnt >= 50
GROUP BY author_name
ORDER BY avg_rating DESC
LIMIT 10
'''
author_highest = pd.io.sql.read_sql(author_highest, con = engine)
author_highest

,author_name,avg_rating
0,J.K. Rowling/Mary GrandPré,4.283844
1,Markus Zusak/Cao Xuân Việt Khương,4.264151
2,J.R.R. Tolkien,4.258446
3,Louisa May Alcott,4.192308
4,Rick Riordan,4.080645
5,William Golding,3.901408
6,J.D. Salinger,3.825581
7,Paulo Coelho/Alan R. Clarke/Özdemir İnce,3.789474
8,William Shakespeare/Paul Werstine/Barbara A. M...,3.787879
9,Dan Brown,3.754540


### Вывод:
- автор с самой высокой средней оценкой книг 4,28 - J.K. Rowling/Mary GrandPré - написавшая Гарри Поттер

##   - Посчитайте среднее количество обзоров от пользователей, которые поставили больше 48 оценок.

In [12]:
# вычеслим среднее количество обзоров от пользователей
number_user_reviews = '''
SELECT
AVG(sub.cnt_text)
FROM
(SELECT 
   reviews.username,
   COUNT(DISTINCT reviews.text) AS cnt_text,
   COUNT(DISTINCT ratings.book_id) AS cnt_book
FROM reviews
     INNER JOIN ratings ON ratings.username = reviews.username
GROUP BY
   reviews.username
HAVING
   COUNT(DISTINCT ratings.book_id)>48) AS sub
'''
number_user_reviews = pd.io.sql.read_sql(number_user_reviews, con = engine)
number_user_reviews

,avg
0,24.0


### Вывод:
- среднее количество обзоров от пользователей, которые поставили больше 48 оценок равно 24 обзорам

## Общий вывод:

1. - Посчитайте, сколько книг вышло после 1 января 2000 года;
  - книг вышло после 1 января 2000 года: 821
  
2. - Для каждой книги посчитайте количество обзоров и среднюю оценку;
  - если смотреть по кол-ву обзоров на первом месте по обзорам книга Twilight (Twilight #1) - Сумерки
  - если смотреть по среднему рейтингу то на первом месте книга A Dirty Job (Grim Reaper #1) - Грязная работенка (Мрачный жнец #1)
  
3. - Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;
  - издательство, которое выпустило наибольшее число книг толще 50 страниц - Penguin Books
  
4. - Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;
  - автор с самой высокой средней оценкой книг 4,28 - J.K. Rowling/Mary GrandPré - написавшая Гарри Поттер
  
5. - Посчитайте среднее количество обзоров от пользователей, которые поставили больше 48 оценок.
  - среднее количество обзоров от пользователей, которые поставили больше 48 оценок равно 24 обзорам